### PR TITLE
remote: support directory inputs in runfiles

### PR DESCRIPTION
**Pull request by buchgr**
This change adds support for tree artifacts (directories) in runfiles in remote execution.

### CODE REVIEW

**Comment by buchgr** [comment]
@benjaminp 
I want to add support for input directories in run files for remote execution. I am not entirely sure I know what I am doing - would be great if you had some pointers :-).

File: src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java [hunk 1]
```diff
@@ -119,7 +122,17 @@ void addRunfilesToInputs(
           if (strict && !actionFileCache.getMetadata(localArtifact).getType().isFile()) {
             throw new IOException("Not a file: " + localArtifact.getPath().getPathString());
           }
-          addMapping(inputMap, location, localArtifact);
+
+          if (localArtifact.isTreeArtifact()) {
```
Comment (benjaminp) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java]: Does this actually need a special case? It seems like the old code would simply create a symlink to a tree artifact directory, which should work just as well as symlinking individual files, right?

File: src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java [hunk 2]
```diff
@@ -288,18 +290,18 @@ private static AggregatingArtifactValue createAggregatingValue(
         return null;
       }
       if (inputValue instanceof FileArtifactValue) {
-        inputs.add(Pair.of(input, (FileArtifactValue) inputValue));
+        fileInputs.add(Pair.of(input, (FileArtifactValue) inputValue));
       } else if (inputValue instanceof TreeArtifactValue) {
-        inputs.add(Pair.of(input, ((TreeArtifactValue) inputValue).getSelfData()));
+        directoryInputs.add(Pair.of(input, (TreeArtifactValue) inputValue));
```
Comment (benjaminp) [inline src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java]: Does it work to simply put `((TreeArtifactValue) inputValue).getChildValues()` into the `inputs` list? Then, `AggregatingArtifactValue` wouldn't need to distinguish between directory and normal inputs.

File: src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java [hunk 3]
```diff
@@ -498,16 +498,6 @@ public void expand(Artifact artifact, Collection<? super Artifact> output) {
       Preconditions.checkState(artifact.isMiddlemanArtifact() || artifact.isTreeArtifact(),
           artifact);
       Collection<Artifact> result = expandedInputs.get(artifact);
-
-      // Note that the result can be empty but not null for TreeArtifacts. And it may be null for
-      // non-aggregating middlemen.
-      if (artifact.isTreeArtifact()) {
```
Comment (benjaminp) [inline src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java]: This assertion seems important.

File: src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java [hunk 4]
```diff
@@ -288,18 +290,18 @@ private static AggregatingArtifactValue createAggregatingValue(
...
       } else {
         // We do not recurse in aggregating middleman artifacts.
         Preconditions.checkState(!(inputValue instanceof AggregatingArtifactValue),
             "%s %s %s", artifact, action, inputValue);
       }
     }
     return (action.getActionType() == MiddlemanType.AGGREGATING_MIDDLEMAN)
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java]: There is probably a Very Good Reason why runfiles middlemen are treated differently from other middlemen, but I can't tell right now and even if that's not necesssary, this change is definitely not the place to fix that. `ActionExecutionFunction.java:694` tells *what* we do with it (we don't put them into `expandedArtifacts`) but it doesn't tell *why* :(

Comment (benjaminp) [reply src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java]: The reasons are explained in the commit message of c868c47eaa1.

File: src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java (hunk 2 above)
Comment (benjaminp) [reply src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java]: Yes, I guess it's unavoidable.

**Comment by buchgr** [comment]
+ @ulfjack

File: src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java (hunk 1 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java]: You're right that that would work for local execution, but it wouldn't work for remote execution.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java [hunk 5]
```diff
@@ -119,7 +122,17 @@ void addRunfilesToInputs(
...
+            List<ActionInput> expandedInputs =
+                ActionInputHelper.expandArtifacts(Collections.singletonList(localArtifact),
+                artifactExpander);
+            for (ActionInput input : expandedInputs) {
+              addMapping(inputMap, location.getRelative(((TreeFileArtifact) input).getParentRelativePath()), input);
```
Comment (ulfjack) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java]: Move the exec path computation to a local variable, please.

**Comment by ulfjack** [comment]
My change to make source directories work correctly is here:
https://bazel-review.googlesource.com/c/bazel/+/29872

It doesn't look related to this change.

(It may be broken - I didn't rebase for several months, and there were some conflicts when I rebased it today.)

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 6]
```diff
@@ -528,6 +528,71 @@ EOF
...
+        "dir":  attr.label(mandatory = True),
+    },
+)
+EOF
+
+     bazel test \
+           --spawn_strategy=remote \
+           --remote_executor=localhost:${worker_port} \
+           //a:skylark_output_dir_test \
+           || fail "Failed to run //a:skylark_output_dir_test with remote execution"
```
Comment (ola-rozenfeld) [inline src/test/shell/bazel/remote/remote_execution_test.sh]: Do I understand correctly that this is supposed to print "hello world"? If so, can you please verify that it does?

**Comment by buchgr** [comment]
merged as d4d3d506f4cf6cfaafaeeb717d681ff7784e2384

**Pull request closed without merging**